devito/function.py

```diff
@@ -623,6 +658,91 @@ def symbolic_shape(self):
         ret = tuple(Add(i, -j) for i, j in zip(symbolic_shape, self.staggered))
         return EnrichedTuple(*ret, getters=self.dimensions)
 
+    _C_structname = 'dataobj'
+    _C_typename = 'struct %s *' % _C_structname
+    _C_field_data = 'data'
+    _C_field_size = 'size'
+    _C_field_nopad_size = 'npsize'
+    _C_field_halo_size = 'hsize'
+    _C_field_halo_ofs = 'hofs'
+    _C_field_owned_ofs = 'oofs'
+
+    _C_typedecl = Struct(_C_structname,
+                         [Value('%srestrict' % ctypes_to_cstr(c_void_p), _C_field_data),
+                          Value(ctypes_to_cstr(POINTER(c_int)), _C_field_size),
+                          Value(ctypes_to_cstr(POINTER(c_int)), _C_field_nopad_size),
+                          Value(ctypes_to_cstr(POINTER(c_int)), _C_field_halo_size),
+                          Value(ctypes_to_cstr(POINTER(c_int)), _C_field_halo_ofs),
+                          Value(ctypes_to_cstr(POINTER(c_int)), _C_field_owned_ofs)])
+
+    _C_ctype = POINTER(type(_C_structname, (Structure,),
+                            {'_fields_': [(_C_field_data, c_void_p),
+                                          (_C_field_size, POINTER(c_int)),
+                                          (_C_field_nopad_size, POINTER(c_int)),
+                                          (_C_field_halo_size, POINTER(c_int)),
+                                          (_C_field_halo_ofs, POINTER(c_int)),
+                                          (_C_field_owned_ofs, POINTER(c_int))]}))
+
+    def _C_make_dataobj(self, data):
+        dataobj = byref(self._C_ctype._type_())
+        dataobj._obj.data = data.ctypes.data_as(c_void_p)
+        dataobj._obj.size = (c_int*self.ndim)(*data.shape)
+        # MPI-related fields
+        dataobj._obj.npsize = (c_int*self.ndim)(*[i - sum(j) for i, j in
+                                                  zip(data.shape, self._extent_padding)])
+        dataobj._obj.hsize = (c_int*(self.ndim*2))(*flatten(self._extent_halo))
+        dataobj._obj.hofs = (c_int*(self.ndim*2))(*flatten(self._offset_halo))
+        dataobj._obj.oofs = (c_int*(self.ndim*2))(*flatten(self._offset_owned))
+        return dataobj
+
+    def _C_as_ndarray(self, dataobj):
+        """Cast the data carried by a TensorFunction dataobj to an ndarray."""
```

Comment:
a ndarray

Reply:
no, when it comes to acronyms, the 'n' goes with the sound of the following letter, rather than depending on whether it's a vowel or a consonant . 

a PDE
a DSL
an FEM 
an NDarray

etc